### PR TITLE
Casting columns as a different data type on select, insert and update

### PR DESCRIPTION
**Pull request by billy1624**
## PR Info

- Related https://github.com/SeaQL/seaography/issues/104

## New Features

- [x] Added `ColumnTrait::select_as()` and `ColumnTrait::select_enum_as()` methods: casting select expression of a column
- [x] Added `ColumnTrait::save_as()` and `ColumnTrait::save_enum_as()` methods: casting value of a column in insert and update statement
- [x] Added `select_as` and `save_as` macro attribute to `DeriveEntityModel`

### CODE REVIEW

**Review by karatakis — APPROVED** [review]
Looks fine by me

File: src/entity/column.rs [hunk 1]
```diff
@@ -338,6 +340,42 @@ pub trait ColumnTrait: IdenStatic + Iterable + FromStr {
     fn into_expr(self) -> Expr {
         Expr::expr(self.into_simple_expr())
     }
+
+    /// Cast column expression used in select statement.
+    /// By default it only cast database enum as text.
```
Comment (tyt2y3) [inline src/entity/column.rs:345]: What I don't understand is, under what circumstance is not `default`?
How do I customize the behaviour?
This seems to unfortunately be part of the public API, and is it intended? Can we hide it instead?

Comment (billy1624) [reply src/entity/column.rs:345]: It's intended to be part of the public API. For example, user can override it and cast `citext` column type as `text` during select, `SELECT CAST("citext_column" AS text), ...`.

Comment (billy1624) [reply src/entity/column.rs:345]: I'll add a self-explanatory test case now

Comment (billy1624) [reply src/entity/column.rs:345]: Added https://github.com/SeaQL/sea-orm/pull/1304/commits/6295fcf185532033b8306e179a4e396cb5526461

**Review by tyt2y3 — CHANGES_REQUESTED** [review]
Ah okay, the only remaining question now is naming.

For example `cast_value_enum` only do any casting *only if it is an enum* and so I think it will be better named as `may_be_cast_value_enum`

But then, `cast_value` is also a bit confusing. I understand now it's the counterpart to `cast_select`, but of course `cast_insert_or_update` is too long as a name. So may be we can use the word, `save`.

And so, the entire naming could be: `select_as` and `save_as`, `cast_save_as` etc?

**Comment by billy1624** [comment]
I think `select_as` and `save_as` are short and concise. How about?

- `select_as`: casting columns on select
- `select_enum_as`: casting enum columns on select
- `save_as`: casting columns on insert and update
- `save_enum_as`: casting enum columns on insert and update

**Comment by tyt2y3** [comment]
sounds good!

**Comment by billy1624** [comment]
Nice! I'll work on that :)

**Review by tyt2y3 — APPROVED** [review]

**Comment by praveenperera** [comment]
@billy1624 this does not work on insert:

```
"Query Error: error occurred while decoding column \"email\": mismatched types; Rust type `core::option::Option<alloc::string::String>` (as SQL type `TEXT`) is not compatible with SQL type `citext`"
```

```
#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "users")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    #[serde(skip_deserializing)] // Skip deserializing
    pub id: Uuid,
    pub first_name: String,
    pub last_name: String,
    #[sea_orm(unique)]
    pub nickname: String,
    pub password: String,
    #[sea_orm(
        column_type = r#"Custom("citext".into())"#,
        select_as = "text",
        save_as = "citext"
    )]
    pub email: String,
    pub country: String,
    #[serde(skip_deserializing)]
    pub created_at: DateTimeUtc,
    #[serde(skip_deserializing)]
    pub updated_at: DateTimeUtc,
}
```

**Comment by praveenperera** [comment]
@billy1624 Hey this does not seem to work, I tried the sea-orm on master:

```
sea-orm = {git = "https://github.com/SeaQL/sea-orm", features = ["runtime-tokio-native-tls", "sqlx-postgres", "macros"]}
```

Tried inserting:

```rust
    let user_params = UserActiveModel::from_json(user_json)?.set_timestamps();
    let user = user_params.insert(&state.database_connection).await?;
```

Entity:

```rust
//! `SeaORM` Entity. Generated by sea-orm-codegen 0.10.7
use sea_orm::{entity::prelude::*, Set};
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, PartialEq, DeriveEntityModel, Eq, Serialize, Deserialize)]
#[sea_orm(table_name = "users")]
pub struct Model {
    #[sea_orm(primary_key, auto_increment = false)]
    #[serde(skip_deserializing)] // Skip deserializing
    pub id: Uuid,
    pub first_name: String,
    pub last_name: String,
    #[sea_orm(unique)]
    pub nickname: String,
    pub password: String,
    #[sea_orm(
        column_type = r#"Custom("citext".into())"#,
        save_as = "citext"
        select_as = "citext",
    )]
    pub email: String,
    pub country: String,
    #[serde(skip_deserializing)]
    pub created_at: DateTimeUtc,
    #[serde(skip_deserializing)]
    pub updated_at: DateTimeUtc,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
pub enum Relation {}

impl ActiveModelBehavior for ActiveModel {}

impl ActiveModel {
    pub fn set_timestamps(mut self) -> Self {
        if self.created_at.is_not_set() {
            self.created_at = Set(chrono::Utc::now())
        }
        self.updated_at = Set(chrono::Utc::now());
        self
    }
}
```

Error

```
Query Error: error occurred while decoding column \"email\": mismatched types; Rust type `core::option::Option<alloc::string::String>` (as SQL type `TEXT`) is not compatible with SQL type `citext
```

**Comment by billy1624** [comment]
Hey @praveenperera, thanks for asking!! Notice the `select_as` is `text`.

```diff
    #[sea_orm(
        column_type = r#"Custom("citext".into())"#,
        save_as = "citext"
-       select_as = "citext",
+       select_as = "text",
    )]
    pub email: String,
```

**Comment by praveenperera** [comment]
Hey @billy1624, sorry that's what I had before and I was changing stuff to try stuff out. I tried select_as ="text" again and same error. I invited you to the repo to check it out. On the `citext` branch: https://github.com/praveenperera/rest-template-rust/tree/citext

**Comment by billy1624** [comment]
Okay. I know what have happened... "a missing `,`" I was surprised that Rust compiler didn't thrown any errors loll

```rs
    #[sea_orm(
        column_type = r#"Custom("citext".into())"#,
-       save_as = "citext"
+       save_as = "citext",
        select_as = "text",
    )]
    pub email: String,
```

**Comment by praveenperera** [comment]
Hmm interesting that actually fixed it. Thanks!

**Comment by billy1624** [comment]
Nice!

**Comment by praveenperera** [comment]
@billy1624 Would this be possible to use when using the `--expanded-format`?

**Pull request merged**